### PR TITLE
Introduce random fleet placement, fix bugs

### PR DESCRIPTION
**Pull request by korotkevics**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally prior to submission?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: client/src/main/java/pl/korotkevics/ships/client/client/OpponentShotTrigger.java [hunk 1]
```diff
@@ -17,9 +18,9 @@
   private static final Target logger = new SharedLogger(Client.class);
 
   @Override
-  public void fire(final Button button, final String messageStatement) {
+  public void fire(final Button button, final Message message) {
```
Comment (Ozymadias) [inline client/src/main/java/pl/korotkevics/ships/client/client/OpponentShotTrigger.java]: why there are at least three such methods in your code?

File: client/src/main/java/pl/korotkevics/ships/client/gui/controllers/FleetPlacementController.java [hunk 2]
```diff
@@ -371,4 +389,34 @@ private void loadGameWindow() {
...
+        forEach(i -> this.drawMast(this.convertToGridIndex(i)));
+  }
+
+  private void drawMast(int index) {
+    logger.info("draw index: " + index);
+    Rectangle rec = (Rectangle) (yourBoard.getChildren().get(index));
+    rec.setFill(Color.GREEN);
+  }
+
+  private int convertToGridIndex(final int index) {
```
Comment (Ozymadias) [inline client/src/main/java/pl/korotkevics/ships/client/gui/controllers/FleetPlacementController.java]: why exactly same method as in GameController?

File: client/src/main/java/pl/korotkevics/ships/client/gui/controllers/FleetPlacementController.java [hunk 3]
```diff
@@ -371,4 +389,34 @@ private void loadGameWindow() {
...
+    this.loadGameWindow();
+  }
+
+  private void drawFleet(final Fleet fleet) {
+    fleet.
+        toIntegerList().
+        forEach(i -> this.drawMast(this.convertToGridIndex(i)));
+  }
+
+  private void drawMast(int index) {
```
Comment (Ozymadias) [inline client/src/main/java/pl/korotkevics/ships/client/gui/controllers/FleetPlacementController.java]: why no final in this case?

File: client/src/main/java/pl/korotkevics/ships/client/gui/controllers/GameController.java [hunk 4]
```diff
@@ -186,30 +186,32 @@ private void loadWithdrawScreen() {
...
+    final int row = shotIndex - (column * BOARD_SIZE);
+    return row * BOARD_SIZE + column + 1;
+  }
+
   private void changeTurn() {
     final double opacity = 0.4;
-    opponentBoard.setDisable(true);
-    opponentBoard.setOpacity(opacity);
+    this.opponentBoard.setDisable(true);
+    this.opponentBoard.setOpacity(opacity);
```
Comment (Ozymadias) [inline client/src/main/java/pl/korotkevics/ships/client/gui/controllers/GameController.java]: why did you add this?

File: client/src/main/java/pl/korotkevics/ships/client/validators/ShipPlacementValidator.java [hunk 5]
```diff
@@ -140,12 +140,12 @@ private boolean isShipAboveShipHorizontal() {
...
   private boolean isShipOnTheRightHorizontal() {
-    return shipStartIndex + mastCount * BOARD_SIZE < board.length
-        && board[shipStartIndex + mastCount * BOARD_SIZE].equals(FieldState.OCCUPIED);
+    return this.shipStartIndex + this.mastCount * BOARD_SIZE < this.board.length
+        && board[this.shipStartIndex + this.mastCount * BOARD_SIZE].equals(FieldState.OCCUPIED);
   }
 
   private boolean isShipOnTheLeftHorizontal() {
-    return shipStartIndex - BOARD_SIZE >= 1 && board[shipStartIndex - BOARD_SIZE].equals(FieldState.OCCUPIED);
+    return this.shipStartIndex - BOARD_SIZE >= 1 && this.board[shipStartIndex - BOARD_SIZE].equals(FieldState.OCCUPIED);
```
Comment (Ozymadias) [inline client/src/main/java/pl/korotkevics/ships/client/validators/ShipPlacementValidator.java]: why did you add "this" in many places?

File: server/src/main/java/pl/korotkevics/ships/server/gamestates/fleetplacement/FleetGenerator.java [hunk 6]
```diff
@@ -16,47 +23,60 @@
...
+      shipAsIndices = this.offerIndices(shipLength, this.generateStartIndex(), BOARD_WIDTH);
+      if (!this.isWithinBoardVertical(shipAsIndices)) {
+        return this.offerShipAsIndices(shipLength);
+      }
     } else {
-      shipAsIndices = this.offerIndices(shipLength, startIndex, 1);
+      shipAsIndices = this.offerIndices(shipLength, this.generateStartIndex(), 1);
+      if (!this.isWithinBoardAndSingleRow(shipAsIndices)) {
+        return this.offerShipAsIndices(shipLength);
+      }
```
Comment (Ozymadias) [inline server/src/main/java/pl/korotkevics/ships/server/gamestates/fleetplacement/FleetGenerator.java]: once again why some many "this" keywords

File: server/src/main/java/pl/korotkevics/ships/server/gamestates/fleetplacement/FleetGenerator.java [hunk 7]
```diff
@@ -16,47 +23,60 @@
...
     }
     return shipAsIndices;
   }
 
-  private List<Integer> offerIndices(int length, int startIndex, int step) {
-    List<Integer> indices = new ArrayList<>();
-    IntStream.range(0, length).forEach(i -> indices.add(startIndex + i * step));
+  private List<Integer> offerIndices(final int length,final int startIndex,final int step) {
```
Comment (Ozymadias) [inline server/src/main/java/pl/korotkevics/ships/server/gamestates/fleetplacement/FleetGenerator.java]: formatting as i guess (spaces after comas)

File: server/src/main/java/pl/korotkevics/ships/server/gamestates/fleetplacement/FleetGenerator.java [hunk 8]
```diff
@@ -76,7 +96,14 @@ private void markAreaAsOccupied(final List<Integer> offeredShip) {
         .forEach(i -> gameBoard.put(i, FieldState.OCCUPIED));
   }
 
-  private boolean isWithinBoard(final List<Integer> offeredShip) {
+  private boolean isWithinBoardAndSingleRow(final List<Integer> offeredShip) {
```
Comment (Ozymadias) [inline server/src/main/java/pl/korotkevics/ships/server/gamestates/fleetplacement/FleetGenerator.java]: a little bit strange name of method

File: server/src/main/java/pl/korotkevics/ships/server/gamestates/fleetplacement/FleetGenerator.java [hunk 9]
```diff
@@ -76,7 +96,14 @@ private void markAreaAsOccupied(final List<Integer> offeredShip) {
...
+    final int rowNumberOfFirstShipIndex = offeredShip.get(0) / BOARD_WIDTH;
+    return offeredShip
+        .stream()
+        .allMatch(i -> i < 100 && (i / BOARD_WIDTH == rowNumberOfFirstShipIndex));
```
Comment (Ozymadias) [inline server/src/main/java/pl/korotkevics/ships/server/gamestates/fleetplacement/FleetGenerator.java]: why in this case you doesn't check whether i > 0 and in method below you does; also why this case you check whether row is same and in method below you doesn't check whether column is same

File: server/src/main/java/pl/korotkevics/ships/server/gamestates/play/PlayState.java [hunk 10]
```diff
@@ -67,7 +67,18 @@ public GameState process() {
   }
 
   private void sendYourTurnMessage() {
+    this.rest();
     MessageSender messageSender = new MessageSender(communicationBus, logger);
-    messageSender.send(turnManager.getCurrentPlayer(), Header.YOUR_TURN);
+    messageSender.send(this.turnManager.getCurrentPlayer(), Header.YOUR_TURN);
```
Comment (Ozymadias) [inline server/src/main/java/pl/korotkevics/ships/server/gamestates/play/PlayState.java]: again why "this"?

File: shared/src/main/java/pl/korotkevics/ships/shared/fleet/Mast.java [hunk 11]
```diff
@@ -13,11 +13,26 @@
...
+   *
+   * @param index location of a mast within a game board grid
+   * @return Mast
+   */
+  public static Mast ofIndex(final String index) {
     return new Mast(index);
   }
+  
+  /**
+   * It is not a subject to change.
```
Comment (Ozymadias) [inline shared/src/main/java/pl/korotkevics/ships/shared/fleet/Mast.java]: I don't understand this information

File: shared/src/test/java/pl/korotkevics/ships/shared/fleet/FleetTest.java [hunk 12]
```diff
@@ -48,9 +49,16 @@ public void stringRepresentationIsProper() {
...
+  }
+  
+  @Test
+  public void integerListRepresentationIsProper() {
+    //given
+    Fleet fleet = Fleet.ofShips(Arrays.asList(Ship.ofMasts(Mast.ofIndex("3"), Mast.ofIndex("2"), Mast.ofIndex("1"))));
+    //when
+    List<Integer> list = fleet.toIntegerList();
+    //then
+    assertEquals(list.toString(), "[1, 2, 3]");
```
Comment (Ozymadias) [inline shared/src/test/java/pl/korotkevics/ships/shared/fleet/FleetTest.java]: strange method of testing collection. I think you should change it

File: client/src/main/java/pl/korotkevics/ships/client/client/OpponentShotTrigger.java (hunk 1 above)
Comment (M-Aarsman) [reply client/src/main/java/pl/korotkevics/ships/client/client/OpponentShotTrigger.java]: all classes  implements one interface :)
